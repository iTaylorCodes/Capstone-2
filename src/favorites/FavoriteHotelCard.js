import { useEffect, useState } from "react";
import { PlacesApi } from "../api/api";
import FavoriteHeart from "../favorites/FavoriteHeart";

/** Favorite Hotel Card.
 *
 * Component to show a card with hotel details provided by HotelsAPI specific to use favorites.
 *
 * Renders -> Hotel ID, Name, Image, Star Rating, Neighborhood, and Address.
 *  + FavoriteHeart to add/remove hotel from user favorites.
 *
 * Rendered by FavoritesList.
 */
const FavoriteHotelCard = ({ hotel }) => {
  const starIcons = [];
  const [hotelImage, setHotelImage] = useState("");

  for (let s = 1; s <= hotel.propertyDescription.starRating; s++) {
    starIcons.push(s);
  }

  const hotelAddress = `${hotel.propertyDescription.address.fullAddress}`;

  useEffect(() => {
    let isSubscribed = true;
    const fetchImage = async () => {
      try {
        const image = await PlacesApi.getImage(
          `${hotel.propertyDescription.name}, ${hotel.propertyDescription.address.cityName}`
        );
        if (isSubscribed) setHotelImage(image);
      } catch (e) {
        console.error(e);
      }
    };
    fetchImage();
    return () => (isSubscribed = false);
  }, [
    hotel.propertyDescription.name,
    hotel.propertyDescription.address.cityName,
  ]);

  return (
    <div className="HotelCard m-2">
      <div className="card" style={{ maxWidth: "435px" }}>
        <img
          src={hotelImage}
          alt={hotel.propertyDescription.name}
          style={{ maxHeight: "300px" }}
        />
        <h3 className="card-header">{hotel.propertyDescription.name}</h3>
        <div className="card-body">
          <FavoriteHeart hotelId={hotel.pdpHeader.hotelId} />
          <h4 className="card-title">
            Rating:{" "}
            {starIcons.map((s) => (
              <i className="fa fa-star" key={s} style={{ color: "gold" }}></i>
            ))}
          </h4>
          <p className="card-text">
            <span className="fw-bold">Neighborhood: </span>
            {hotel.propertyDescription.address.cityName}
          </p>
          <p className="card-text">
            <span className="fw-bold">Address: </span>
            {hotelAddress}
          </p>
        </div>
      </div>
    </div>
  );
};

export default FavoriteHotelCard;
