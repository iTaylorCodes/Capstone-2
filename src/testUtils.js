import UserContext from "./auth/UserContext";

const testUser = {
  username: "testuser",
  first_name: "testfirst",
  last_name: "testlast",
  email: "test@test.net",
  password: "password",
  favoritedProperties: ["12345"],
};

const UserProvider = ({ children, currentUser = testUser }) => (
  <UserContext.Provider value={{ currentUser }}>
    {children}
  </UserContext.Provider>
);

export { UserProvider };
