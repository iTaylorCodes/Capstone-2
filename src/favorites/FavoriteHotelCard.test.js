import FavoriteHotelCard from "./FavoriteHotelCard";
import { render } from "@testing-library/react";
import { UserProvider } from "../testUtils";

const hotel = {
  pdpHeader: { hotelId: "12345" },
  propertyDescription: {
    name: "Tester Hotel",
    address: {
      cityName: "Los Angeles",
      fullAddress: "1234 Tester Way, Los Angeles, CA 12345",
    },
    starRating: "5",
  },
};

it("renders without crashing", () => {
  render(
    <UserProvider>
      <FavoriteHotelCard hotel={hotel} />
    </UserProvider>
  );
});

it("matches snapshot", () => {
  const { asFragment } = render(
    <UserProvider>
      <FavoriteHotelCard hotel={hotel} />
    </UserProvider>
  );
  expect(asFragment()).toMatchSnapshot();
});
