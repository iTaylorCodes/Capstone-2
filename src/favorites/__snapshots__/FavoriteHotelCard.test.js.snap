// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="HotelCard m-2"
  >
    <div
      class="card"
      style="max-width: 435px;"
    >
      <img
        alt="Tester Hotel"
        src=""
        style="max-height: 300px;"
      />
      <h3
        class="card-header"
      >
        Tester Hotel
      </h3>
      <div
        class="card-body"
      >
        <div
          class="FavoriteHeart d-flex justify-content-end"
          style="margin: -5px;"
        >
          <i
            class="fa fa-heart"
            data-testid="favHeart"
            style="color: rgb(2, 117, 216); cursor: pointer;"
          />
        </div>
        <h4
          class="card-title"
        >
          Rating: 
          <i
            class="fa fa-star"
            style="color: gold;"
          />
          <i
            class="fa fa-star"
            style="color: gold;"
          />
          <i
            class="fa fa-star"
            style="color: gold;"
          />
          <i
            class="fa fa-star"
            style="color: gold;"
          />
          <i
            class="fa fa-star"
            style="color: gold;"
          />
        </h4>
        <p
          class="card-text"
        >
          <span
            class="fw-bold"
          >
            Neighborhood: 
          </span>
          Los Angeles
        </p>
        <p
          class="card-text"
        >
          <span
            class="fw-bold"
          >
            Address: 
          </span>
          1234 Tester Way, Los Angeles, CA 12345
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
